kube-core/src/labels.rs

```diff
@@ -0,0 +1,552 @@
+            .unwrap_or_default();
+        equality.0.extend(expressions);
+        equality
+    }
+}
+
+impl From<LabelSelectorRequirement> for Expression {
+    fn from(requirement: LabelSelectorRequirement) -> Self {
+        let key = requirement.key;
+        let values = requirement.values.map(|values| values.into_iter().collect());
+        match requirement.operator.as_str() {
+            "In" => match values {
+                Some(values) => Expression::In(key, values),
+                None => Expression::Invalid,
+            },
+            "NotIn" => match values {
+                Some(values) => Expression::NotIn(key, values),
+                None => Expression::Invalid,
+            },
+            "Exists" => Expression::Exists(key),
+            "DoesNotExist" => Expression::DoesNotExist(key),
+            _ => Expression::Invalid,
+        }
+    }
+}
+
+impl From<Selector> for LabelSelector {
+    fn from(value: Selector) -> Self {
+        let mut equality = vec![];
+        let mut expressions = vec![];
+        for expr in value.0 {
+            match expr {
+                Expression::In(key, values) => expressions.push(LabelSelectorRequirement {
+                    key,
+                    operator: "In".into(),
+                    values: Some(values.into_iter().collect()),
+                }),
+                Expression::NotIn(key, values) => expressions.push(LabelSelectorRequirement {
+                    key,
+                    operator: "NotIn".into(),
+                    values: Some(values.into_iter().collect()),
+                }),
+                Expression::Equal(key, value) => equality.push((key, value)),
+                Expression::NotEqual(key, value) => expressions.push(LabelSelectorRequirement {
+                    key,
+                    operator: "NotIn".into(),
+                    values: Some(vec![value]),
+                }),
+                Expression::Exists(key) => expressions.push(LabelSelectorRequirement {
+                    key,
+                    operator: "Exists".into(),
+                    values: None,
+                }),
+                Expression::DoesNotExist(key) => expressions.push(LabelSelectorRequirement {
+                    key,
+                    operator: "DoesNotExist".into(),
+                    values: None,
+                }),
+                Expression::Invalid => (),
+            }
+        }
+
+        LabelSelector {
+            match_labels: (!equality.is_empty()).then_some(equality.into_iter().collect()),
+            match_expressions: (!expressions.is_empty()).then_some(expressions),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use std::iter::FromIterator;
+
+    #[test]
+    fn test_raw_matches() {
+        for (selector, labels, matches, msg) in &[
+            (Selector::default(), Default::default(), true, "empty match"),
+            (
+                Selector::from_iter(Some(("foo", "bar"))),
```

Comment:
Selector from_iter `Option<Pair>`.. what does that actually evaluate to? Is that two `Exist` expressions?

Reply:
Added a doctest which gives some examples. It is the one which goes to `from_map` and to `Equality` type of expression.